packages/node_modules/@node-red/nodes/core/network/10-mqtt.js

```diff
@@ -837,7 +837,7 @@ module.exports = function(RED) {
                         resolve();
                      } else {
                          const t = setTimeout(reject, ms);
-                         client.end(() => {
+                         client.end(true, () => {
```

Comment:
I propose rather than blindly force end, we at least attempt a clean end first (so that any in-transit messages are ack'd). 

What are your thoughts on something like this...

```js
        node.disconnect = function (callback) {
            const WAIT_END = 1000;
            const _callback = function () {
                if(node.connected || node.connecting) {
                    setStatusDisconnected(node, true);
                }
                if(node.client) { node.client.removeAllListeners(); }
                node.connecting = false;
                node.connected = false;
                callback && typeof callback == "function" && callback();
            };
            if(!node.client) { return _callback(); }
            if(node.closing) { return _callback(); }

            let waitEnd = (client, ms) => {
                return new Promise( (resolve, reject) => {
                    node.closing = true;
                    if(!client) { 
                        resolve();
                     } else {
                         const t = setTimeout(() => {
                             //clean end() has exceeded WAIT_END, lets force end!
                            client && client.end(true);
                            reject();
                         }, ms);
                         client.end(() => {
                             clearTimeout(t);
                             resolve();
                         });
                     }
                });
            };
            if(node.connected && node.closeMessage) {
                node.publish(node.closeMessage, function (err) {
                    waitEnd(node.client, WAIT_END).then(() => {
                        _callback();
                    }).catch((e) => {
                        _callback();
                    })
                });
            } else {
                waitEnd(node.client, WAIT_END).then(() => {
                    _callback();
                }).catch((e) => {
                    _callback();
                })
            }
        }
```

Reply:
Hi Steve,  I like the idea , but when I tried in in my test system even though I can see the second call to client.end() being made from the timeout it didn't clear the ping timer in the mqtt library.  I think maybe the first call to client.end() has done enough of a clean up to convince the mqtt library that the second call (with the force option) becomes a no-op.

Reply:
I'll dig a bit more into the mqtt layer to try and understand what's happening - it seems like a ping timer gets left running if you try to send a message to a mqtt server which has stopped.

Reply:
I've tracked the root cause down now and updated the issue to reflect this.   The problem was due to the nod-red code calling removeAllListeners() and removing an event handler that the mqtt client itself depends on.

This new PR now tracks and removes just the handlers added by the node-red code